### ISSUE TITLE
[Flatpak] Call for native file chooser

### ISSUE THREAD

**Issue by Doomsdayrs** [ATLauncher/ATLauncher#564]
**Is your feature request related to a problem? Please describe.**
Flatpak versions file picker does not show files outside of the sandbox.

**Describe the solution you'd like**
Call for the system file picker, which will provide better support when sandboxed.

**Additional context**
Flatpak version has a filesystem sandbox. 
Because of this, the file picker will only show the sandboxed view.

**Comment by RyanTheAllmighty** [comment]
Any examples of code or documentation to go along with this?

**Comment by Doomsdayrs** [comment]
I'm gathering the documentation.

While I do that.
Can you point me to where in ATLaunchers code calls for the file pciker?

**Comment by RyanTheAllmighty** [comment]
Have you tried enabling the use native file picker option in the settings?

**Comment by Doomsdayrs** [comment]
> Have you tried enabling the use native file picker option in the settings?

Yes, Both leads to a sandboxed view.

**Comment by Doomsdayrs** [comment]
@RyanTheAllmighty the reason I ask for where the code is located is to understand what solution is needed.

For example, with GTK the following is the solution:
https://docs.gtk.org/gtk4/class.FileChooserNative.html

**Comment by Doomsdayrs** [comment]
The underlying approach is using [portals](https://github.com/flatpak/xdg-desktop-portal)

For example, the specific portal for this is [FileChooser](https://flatpak.github.io/xdg-desktop-portal/#gdbus-org.freedesktop.portal.FileChooser)

**Comment by RyanTheAllmighty** [comment]
It's in https://github.com/ATLauncher/ATLauncher/blob/master/src/main/java/com/atlauncher/gui/dialogs/FileChooserDialog.java

AFAIK it's already using the native one that comes with Swing because there were issues with OSX sandboxing that required using the JFileChooser which, from memory, uses native OS picker instead of a Swing based gui

**Comment by Doomsdayrs** [comment]
Then it is expected that the JFileChooser not be able to display files outside of the sandbox, since it is working with java tools that will be sandboxed as well.

I'll have to configure the flatpak to have the native file chooser enabled by default.

But building upon that, where is the call for the native file chooser (when enabled)? Since, while it does call a native dialog, is still sandboxed.

**Comment by Doomsdayrs** [comment]
Or am I misunderstanding something and Swing itself calls for a native dialog?

**Comment by RyanTheAllmighty** [comment]
Yeah the file I posted before checks if the native dialog setting is enabled, and of so uses the JFileChooser which is a Java component which uses a native dialog under the hood. It's all hidden away inside the JRE and that file is the highest level call we have. So if that doesn't work, then likely there's something else that needs to be done

**Comment by Doomsdayrs** [comment]
Alright then, if that is the case. We might need to call the portals API directly if it exists.

I can create the JNI library for this

**Comment by Doomsdayrs** [comment]
What must be used then is the following

1. https://libportal.org/method.Portal.open_file.html
2. https://libportal.org/method.Portal.open_file_finish.html

**Comment by Doomsdayrs** [comment]
Looking into both JNI & JNA, both seem to require far more effort to accomplish.

**Comment by Doomsdayrs** [comment]
The equally horrible option is a DBUS call directly with xdg-desktop-portal

**Comment by Doomsdayrs** [comment]
I decided the best solution for this is to push the issue onto Java itself by contacting Oracle requesting utilization of FreeDesktop portals for the FileDialog.

Internal review ID: 9072984

**Comment by RyanTheAllmighty** [comment]
It's very likely something we can do and fix. They're would be other Java programs that would have file pickers that must've solved the issue already. IDEs like intellij or eclipse for example

**Comment by Doomsdayrs** [comment]
The IDEs just get full access to all files.

**Comment by Doomsdayrs** [comment]
Awesome! I'll compile to see how it works under flatpak.

Regarding jar size, I'll quickly work on implementing proguard.

**Comment by Doomsdayrs** [comment]
Here I added the foundations for proguard to shrink the jar size 463f0f1e0853d8ef083eae2986c8ccd898da1510

**Comment by Doomsdayrs** [comment]
Note, `FileSystem.BASE_DIR` turns into a string of `**USERSDIR**`

**Comment by Doomsdayrs** [comment]
Alright, now that files are being selected, the issue comes down to the returned URI being parsed incorrectly.

Some interesting logs are the following. It seems that it is trying to append the uri provided to the BASE_DIR, which would not make sense whatsoever.
```
[05/04/2022 12:36:28 PM] File /home/doomsdayrs/file:/run/user/1000/doc/331f251b/AI-Improvements-1.15.2-0.3.0.jar cannot be copied to **USERSDIR**/instances/AlltheMods7ATM71182/mods/AI-Improvements-1.15.2-0.3.0.jar as it isn't a file
[05/04/2022 12:36:28 PM] File /home/doomsdayrs/file:/run/user/1000/doc/331f251b/AI-Improvements-1.15.2-0.3.0.jar cannot be copied to **USERSDIR**/instances/AlltheMods7ATM71182/mods/AI-Improvements-1.15.2-0.3.0.jar as it doesn't exist
```


Console logs:
```
Unique name: :1.399
       result: /org/freedesktop/portal/desktop/request/1_399/45e810fdacc549dc864eff6306c45148
expected path: /org/freedesktop/portal/desktop/request/1_399/45e810fdacc549dc864eff6306c45148
--- signal received ---
DBusSignal(1,341) { Path=>/org/freedesktop/portal/desktop/request/1_399/45e810fdacc549dc864eff6306c45148, Interface=>org.freedesktop.portal.Request, Member=>Response, Destination=>:1.399, Sender=>:1.88, Signature=>ua{sv} } { 0, { choices => [[]],uris => [[file:///run/user/1000/doc/331f251b/AI-Improvements-1.15.2-0.3.0.jar]] } }
params: size=2
Screenshot successfull
uris: [file:///run/user/1000/doc/331f251b/AI-Improvements-1.15.2-0.3.0.jar]
file:///run/user/1000/doc/331f251b/AI-Improvements-1.15.2-0.3.0.jar
java.nio.file.NoSuchFileException: file:/run/user/1000/doc/331f251b/AI-Improvements-1.15.2-0.3.0.jar
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:86)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
	at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:214)
	at java.nio.file.Files.newByteChannel(Files.java:361)
	at java.nio.file.Files.newByteChannel(Files.java:407)
	at java.nio.file.spi.FileSystemProvider.newInputStream(FileSystemProvider.java:384)
	at java.nio.file.Files.newInputStream(Files.java:152)
	at com.atlauncher.utils.ArchiveUtils.getFile(ArchiveUtils.java:84)
	at com.atlauncher.utils.Utils.getMCModForFile(Utils.java:1597)
	at com.atlauncher.gui.dialogs.EditModsDialog.generateMod(EditModsDialog.java:332)
	at com.atlauncher.gui.dialogs.EditModsDialog.lambda$setupComponents$4(EditModsDialog.java:249)
	at java.lang.Thread.run(Thread.java:748)
java.nio.file.NoSuchFileException: file:/run/user/1000/doc/331f251b/AI-Improvements-1.15.2-0.3.0.jar
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:86)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
	at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:214)
	at java.nio.file.Files.newByteChannel(Files.java:361)
	at java.nio.file.Files.newByteChannel(Files.java:407)
	at java.nio.file.spi.FileSystemProvider.newInputStream(FileSystemProvider.java:384)
	at java.nio.file.Files.newInputStream(Files.java:152)
	at com.atlauncher.utils.ArchiveUtils.getFile(ArchiveUtils.java:84)
	at com.atlauncher.utils.Utils.getFabricModForFile(Utils.java:1611)
	at com.atlauncher.gui.dialogs.EditModsDialog.generateMod(EditModsDialog.java:338)
	at com.atlauncher.gui.dialogs.EditModsDialog.lambda$setupComponents$4(EditModsDialog.java:249)
	at java.lang.Thread.run(Thread.java:748)
```
ATLauncher console logs:
[atlauncher_flatpak-error-1.txt](https://github.com/ATLauncher/ATLauncher/files/8420194/atlauncher_flatpak-error-1.txt)

**Comment by Doomsdayrs** [comment]
TLDR; ATLauncher expected a file system path, not a URI.
I patched the different functions to work with URIs properly.

**Comment by RyanTheAllmighty** [comment]
This has been reverted and reopened.

This change has caused issues with installing 1.16.5 versions of Forge:

```
23:42:51.779 [ATL-Logging-Thread] DEBUG -   Patching net/minecraft/client/Minecraft 1/1
23:42:51.779 [ATL-Logging-Thread] ERROR - java.io.IOException: Patch expected net/minecraft/client/Minecraft to have the checksum 341adb7b but it was 573fdc33
23:42:51.784 [ATL-Logging-Thread] ERROR - 	at net.minecraftforge.binarypatcher.Patcher.patch(Patcher.java:162)
23:42:51.786 [ATL-Logging-Thread] ERROR - 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
23:42:51.786 [ATL-Logging-Thread] ERROR - 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
23:42:51.786 [ATL-Logging-Thread] ERROR - 	at java.base/java.lang.Thread.run(Thread.java:833)
23:42:51.791 [ATL-Logging-Thread] ERROR - java.io.IOException: Patch expected net/minecraft/client/Minecraft to have the checksum 341adb7b but it was 573fdc33
	at net.minecraftforge.binarypatcher.Patcher.patch(Patcher.java:162)
	at net.minecraftforge.binarypatcher.Patcher.process(Patcher.java:114)
	at net.minecraftforge.binarypatcher.ConsoleTool.main(ConsoleTool.java:128)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at com.atlauncher.data.minecraft.loaders.forge.Processor.process(Processor.java:197)
	at com.atlauncher.data.minecraft.loaders.forge.Forge113Loader.lambda$1(Forge113Loader.java:96)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at com.atlauncher.data.minecraft.loaders.forge.Forge113Loader.runProcessors(Forge113Loader.java:93)
	at com.atlauncher.workers.InstanceInstaller.installLoader(InstanceInstaller.java:2060)
	at com.atlauncher.workers.InstanceInstaller.install(InstanceInstaller.java:1330)
	at com.atlauncher.workers.InstanceInstaller.doInBackground(InstanceInstaller.java:344)
	at com.atlauncher.workers.InstanceInstaller.doInBackground(InstanceInstaller.java:1)
	at java.desktop/javax.swing.SwingWorker$1.call(SwingWorker.java:304)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.desktop/javax.swing.SwingWorker.run(SwingWorker.java:343)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

First guess would be something in the dbus librarys deps is messing with the classpath maybe so when it goes to run the Forge processors, something is messing up.

Either way reverting majority of the changes done for this in a single commit that can end up being reverted when a fix is found.

**Comment by Doomsdayrs** [comment]
This doesn't make sense, Why would the patch have a different checksum due to a file selector change?

**Comment by Doomsdayrs** [comment]
I am trying to wrap my head around *what* is going on here

**Comment by Doomsdayrs** [comment]
The only dependency real DBUS uses is  `org.slf4j:slf4j-api:1.7.36`

So it must be JNR

**Comment by Doomsdayrs** [comment]
But JNR is removed, so it must be some other weak dependency??

**Comment by RyanTheAllmighty** [comment]
Yeah not sure why it happens, all I know is that it goes away when the dependencies are removed. I'll have to take a deeper look later on, just busy moving for the next couple weeks so just yanked it out to get a quick fix out

**Comment by Doomsdayrs** [comment]
I finally got around to comparing the two jars... (v3.4.13.1 vs v3.4.13.2)

Nothing interferes with any other part of the code. Except 1 part.
SLF4J has an inconsistency, with classes being removed between the two versions.

But why would a logging issue interfere?

**Comment by Doomsdayrs** [comment]
For note
Forge 1.15.1 works fine
Forge 1.16.# fails
Forge 1.17.1 works fine

**Comment by Doomsdayrs** [comment]
This might be a forge issue, I suppose I will have to make an issue on MinecraftForge's github

**Comment by RyanTheAllmighty** [comment]
Nah this is 100% a launcher issue and nothing related to Forge. Something with the way we run their installer logic and something in the classpath. No need to raise any issues with Forge, until I've managed to have a look and confirm things as I'd be 100% confident it's the launcher

**Comment by Doomsdayrs** [comment]
Alright, I think I hit a eureka moment just before having to go to work rn.

The solution for the issue is simply move the job of calling the DBUS api etc to another micro program.

Then ATLauncher only needs to check for the presence of the extra program to execute it, and retrieve the results from std out.

Genius I must say, I'll work on it later tonight.
